src/webui/api/torrentscontroller.cpp

```diff
@@ -1523,3 +1582,79 @@ void TorrentsController::setSSLParametersAction()
 
     torrent->setSSLParameters(sslParams);
 }
+
+void TorrentsController::metadataAction()
+{
+    const QString source {params()[u"source"_s]};
+
+    if (Net::DownloadManager::hasSupportedScheme(source))
+        throw APIError(APIErrorType::BadParams, tr("Source must be magnet URI or info hash"));
+
+    if (const auto parseResult = BitTorrent::TorrentDescriptor::parse(source))
+    {
+        auto *session = BitTorrent::Session::instance();
+        const BitTorrent::TorrentDescriptor &torrentDescr = parseResult.value();
+        const BitTorrent::InfoHash &infoHash = torrentDescr.infoHash();
+
+        const BitTorrent::Torrent *const torrent = BitTorrent::Session::instance()->findTorrent(infoHash);
+        // torrent has already been added and has metadata
+        if (torrent && torrent->info().isValid())
+        {
+            // the TorrentInfo returned by Torrent::info() contains an empty list of trackers
+            // we must fetch the trackers directly via Torrent::trackers()
+            QVector<BitTorrent::TrackerEntry> trackers;
+            for (const BitTorrent::TrackerEntryStatus &tracker : torrent->trackers())
+                trackers << BitTorrent::TrackerEntry
+                {
+                    .url = tracker.url,
+                    .tier = tracker.tier
+                };
+
+            setResult(serializeTorrentMetadata(torrent->info(), trackers));
+            return;
+        }
+
+        const auto iter = m_torrentMetadata.find(infoHash);
+
+        // metadata has already been downloaded
+        if (iter != m_torrentMetadata.end() && iter.value().isValid())
+        {
+            const BitTorrent::TorrentInfo &metadata = iter.value();
+            setResult(serializeTorrentMetadata(metadata, metadata.trackers()));
+            return;
+        }
+
+        if (!session->isKnownTorrent(infoHash) && iter == m_torrentMetadata.end())
+        {
+            qDebug("Fetching metadata for %s", qUtf8Printable(infoHash.toTorrentID().toString()));
+            if (!session->downloadMetadata(torrentDescr))
+                throw APIError(APIErrorType::BadParams, tr("Unable to download metadata"));
+
+            m_torrentMetadata.insert(infoHash, BitTorrent::TorrentInfo{});
+        }
+
+        setResult(QJsonObject {}, 202);
+        return;
+    }
+
+    throw APIError(APIErrorType::BadParams, tr("Unable to parse torrent"));
+}
+
+void TorrentsController::onMetadataDownloaded(const BitTorrent::TorrentInfo &metadata)
+{
+
+    const auto iter = m_torrentMetadata.find(metadata.infoHash());
+    // only process if a lookup was explicitly initiated via API
+    if (iter != m_torrentMetadata.end())
+    {
+        Q_ASSERT(metadata.isValid());
+        if (!metadata.isValid()) [[unlikely]]
+        {
+            // reattempt the download on the next request
```

Comment:
This is modeled off of `GUIAddTorrentManager::onMetadataDownloaded` which performs this same assertion. According to the [libtorrent docs](https://www.libtorrent.org/reference-Alerts.html#metadata_received_alert), the metadata will be fully received here.

Reply:
>This is modeled off of `GUIAddTorrentManager::onMetadataDownloaded` which performs this same assertion.

Well, then you just misinterpreted it.
Invalid TorrentInfo passed in `GUIAddTorrentManager::onMetadataDownloaded` is a logical error, thus it asserts on `metadata.isValid()`. We do not know the cause of this error and therefore do not know how to recover from it. The conditional expression immediately after assertion serves only to terminate the function rather than trying to recover from error (like you do by "reattempt the download on the next request") since it is meaningless (given the above).
BTW, failed assertion will simply abort the application in many build configurations.

Reply:
Given that we have a recovery path here, I'll remove the assertion.

Reply:
> Given that we have a recovery path here

"Recovery" from what?
By doing this, you allow the caller to pass invalid data to this function, i.e. you believe that there are some expected circumstances when it is acceptable for the caller to do so, right? What do you think these circumstances are?

Reply:
I don't understand - do you want the assertion or not? Invalid data can't be passed here, based on the librorrent alert that triggers this signal. Calling `isValid()` is a good sanity check to have anyway.

Reply:
I want you to remove this meaningless/confusing "recovery" code.

Reply:
You would prefer not to have the `isValid()` sanity check?